config.txt
# Configuration file for MegaZeux 2.93d
# June 9th, 2025

# Comments begin with #. Whitespace is ignored. Empty lines are
# ignored as well.
# Options are set in the following format:
# option = value
# If value is not given, then it is assumed to be 1.
# You can include spaces in values by escaping them with \s.
# (see the default macros for examples)
# For directories, use forward slashes (/) instead of back-slashes,
# as they work more reliably.
# Note that even if the option is given to change the values
# in MZX, this text file will NOT be modified, and hence the
# values will revert to what they are here when you restart MZX.

# NOTE: All of the options below have sane defaults. Changes to
# these options will only take effect if you uncomment them.

# In many cases, the current default value has been written in,
# but this is not guaranteed (for example with video_output).

### Video options ###

# Rendering engine to use to display MegaZeux. The simplest (but not
# necessarily fastest) renderer is "software". Due to MegaZeux's strange
# default resolution (640x350), many video cards have difficulty
# supporting it correctly in fullscreen mode.
#
# The software renderer is the fallback renderer if another renderer fails
# to initialize, and has generally adequate performance. Additionally, it
# supports 8 bit indexed color and requires no overlay or 3D hardware support.
# If you're on a slow computer or a handheld portable, definitely use this.
# To use it, uncomment the line below.

# video_output = software

# A much faster option for almost anything with a graphics card is the
# "softscale" renderer. This renderer uses SDL2's built-in rendering API
# to stream the output of the software renderer to a hardware-accelerated
# texture (if available) for extremely fast scaling and display. This (or
# "glslscale", see below) is usually the fastest scaling renderer for devices
# using integrated graphics e.g. Intel HD, phones, tablets, Raspberry Pi.
#
# Either 32 bit or 16 bit rendering can be enabled (see force_bpp), but
# 32 bit is recommended. Nearest neighbor or linear scaling will be chosen
# automatically based on the scaling ratio and size of the window. This
# renderer is a replacement for the YUV overlay renderer.

# video_output = softscale

# The SDL renderer API used by softscale will use the following underlying
# drivers by default: Direct3D (Windows), Metal (Mac OS 10.11+), or OpenGL.
# To request a different driver, the following option can be changed.
# See https://wiki.libsdl.org/SDL_HINT_RENDER_DRIVER for more info.

# sdl_render_driver = opengl

# If you built MegaZeux yourself with SDL 1.2 support enabled, the YUV overlay
# renderer is available instead of softscale. This renderer is good if you
# want scaling and can't use OpenGL. The quality of scaling is hardware
# dependent and often looks bad at non-integer scaling.
#
# Uncomment either of the lines below (but not both) if you want to try it.
# The "overlay2" mode is cheaper, but will be less accurate due to the use
# of chroma subsampling instead of a full YUV encode. The "overlay2" mode
# also does not support newer graphical features such as unbound sprites.
# If uncommented for SDL2 builds, these lines will enable softscale instead.

# video_output = overlay1
# video_output = overlay2

# An OpenGL renderer is a good choice if you want a configurable, cheap
# renderer and have a video card. This method provides the highest quality
# scaler and should work on all platforms with OpenGL or OpenGL ES support.
#
# MegaZeux provides two fixed-function scalers. The first renders the whole
# screen to a texture and uses OpenGL to stretch it. This method is generally
# fast but may be slow on very old PCs. This renderer is called "opengl1".
#
# The second fixed-function scaler draws entirely in OpenGL by remapping the
# character sets to a texture. This yields a performance increase for some
# games, but at the cost of compatibility. This renderer is called "opengl2".
#
# MegaZeux also provides a scaler for programmable hardware. If you have
# OpenGL 2.0 or OpenGL ES 2.0 capable hardware, you should use this renderer.
# This renderer uses GLSL shaders to render and scale the screen, and several
# different scaling shaders can be selected at runtime using the settings menu.
# This renderer is called "glsl". It is the highest quality and best performing
# renderer for most PCs with a dedicated video card.
#
# A variant of the GLSL renderer called "glslscale" is available that uses the
# software renderer internally with GLSL shader scaling. This performs better
# than the original GLSL renderer on PCs and other devices that don't have a
# video card, and is currently the default renderer for most platforms.
#
# Uncomment ONE of the lines below:

# video_output = opengl1
# video_output = opengl2
# video_output = glsl
# video_output = glslscale

# The scaling renderers can be configured to preserve aspect ratio where
# possible in window and fullscreen modes. The supported ratios are
# "classic" (4:3) and "modern" (64:35). The ratio type "stretch" disables
# any aspect ratio preservation. The default is "modern" (except for consoles
# where "classic" makes more sense, including the 3DS, NDS, and Wii).

# video_ratio = modern

# OpenGL will, by default, linearly interpolate pixels as it upscales
# to your requested resolution. This creates an output with more "blur",
# which some people dislike. If you are using a sufficiently high target
# resolution, preferably an integer multiple of 640x350, you may find
# "nearest" more pleasing, which will have no "blur". The default
# setting is "linear".

# gl_filter_method = nearest

# The GLSL renderer uses scaling shaders to determine the scaling method
# instead of gl_filter_method. The default shader to be loaded when MZX
# starts is assets/glsl/scalers/semisoft.frag.
# Change this config option to specify a different shader to load on startup.
# For a list of valid options, see the F2 menu or the GLSL scalers folder.
# Do not include an extension.

# gl_scaling_shader = semisoft

# Some OpenGL driver settings may default vertical sync on, which looks
# visually better on CRTs but can cause "speed 1" to run noticably slower.
# For consistency with other renderers, and for correctness, MegaZeux will
# attempt to disable vsync when gl_vsync is the default value of 0.
# Conversely, a value of 1 forces MegaZeux to override the driver settings
# and enable vsync, which decreases tearing. A value of -1 can be used to
# tell MegaZeux to obey the driver default, which is unlikely to be useful.

# gl_vsync = 1

# Resolution MZX uses for fullscreen. With the software renderer, this
# will center MZX on the screen, and will not scale it. All other renderers
# listed above will attempt to scale the screen to match this resolution.
# Best results are obtained by using your current desktop resolution or a
# well-supported resolution (like 1280,720) in conjunction with the modern
# aspect ratio.
#
# Give in x,y format. Use this if the defaults are giving you problems,
# especially in the software renderer.
# The current desktop resolution is the default for all scalable renderers;
# "640,480" is the default for the software renderer.

# fullscreen_resolution = 640,480

# If this setting is enabled, instead of regular fullscreen MZX will create
# a borderless window with the current desktop resolution when fullscreen
# is enabled. This overrides fullscreen_resolution. Disabled by default
# except on macOS, where it is enabled by default.

# fullscreen_windowed = 0

# The resolution MZX uses in a window. If your renderer supports
# it, MZX will be scaled to fit this specification. Otherwise
# the setting is ignored.

# window_resolution = 640,350

# Use this to toggle whether MZX can be scaled while in window
# mode, if your renderer supports it. If you use a renderer which
# supports scaling, but don't want to accidentally scale it (which
# may distort the image), you can switch this off here.

# enable_resizing = 1

# MZX is limited internally to 8bit colour, but it can upmix its
# rendering to 32bit if your video card has difficulty with 8bit
# resolutions. Using 32bit colour is slower, but not by much.
#
# NOTE: This option may be ignored by one or more of the above renderers.
# force_bpp is guaranteed to apply to "software". This might not always be
# useful, however; SDL 2 picks a native window pixel format automatically
# and non-native BPPs are handled with an extra surface blit. The "softscale"
# renderer can use this setting to select different native pixel formats.
#
# Valid settings are usually '8', '16', or '32'; a value of '0' or 'auto'
# will instruct the renderer to automatically pick the best available
# BPP (if applicable). Default is 'auto'.

# force_bpp = auto

# Whether MZX should start up in fullscreen or not.
# Press ctrl-alt-enter to toggle fullscreen as MZX runs.

# fullscreen = 1

# Whether MZX should disable the screensaver (or similar) while running.
# By default, MZX will instruct SDL to leave the screensaver on. This setting
# currently only affects SDL 2+ ports.

# disable_screensaver = 0

# MZX can optionally display the text cursor to hint at the currently selected
# UI element when applicable. Valid settings are 0 or "off" (doesn't display
# or update the cursor position), 1 or "hidden" (doesn't display, but updates
# the cursor position for platforms with hardware cursors), "underline"
# (displays the underline cursor), and "solid" (displays the filled cursor).
# The default value is 1.

# dialog_cursor_hints = 1

# Whether MZX should allow a screenshot to be taken when F12 is pressed.
# Defaults to enabled.

# allow_screenshots = 1

### Audio options ###

# Sampling rate to output audio at. Higher values will sound
# better but use more resources and may not be supported by
# your audio card/drivers. Typical values include 22050, 44100,
# and 48000. If your card supports it you might get away with
# 96000, or even 192000.

# audio_sample_rate = 44100

# Audio buffer size in sample frames. Setting this to lower can improve
# latency, so audio will be heard sooner after it is set to be
# changed. It is also possible to change PC speaker effects more
# rapidly with lower values. Lower buffer sizes need more CPU power
# to process. 2048 is adequate for PC speaker playback every
# command at speed 4.

# audio_buffer_samples = 1024

# Number of audio channels. The normal value here is 2 for stereo, but
# in rare cases, 1 (mono) may be preferable. This setting is ignored
# by most ports and for some DOS sound devices. MegaZeux currently
# does not support any type of surround sound.

# audio_output_channels = 2

# Allow music to be sampled at higher precision. Increases CPU
# usage but increases audio quality as well. (Modplug only.)

# enable_oversampling = 0

# Set global resample mode. This affects how OGGs, SAMs, WAVs, RADs, and
# frequency shifted modules sound. Choices are:
# none (fastest, poor quality)
# linear (fast, good quality)
# cubic (slow, great quality)

# resample_mode = linear

# Set resample mode for xmp (or Modplug, depending on build settings).
# This affects how modules sound. Choices are:
# none (fastest, poor quality)
# linear (fast, good quality)
# cubic (slow, great quality)
# fir (very slow, excellent quality) (Modplug only)

# module_resample_mode = cubic

# Whether music/samples should be played or not.
# Does not include PC speaker effects.

# music_on = 1

# Whether PC speaker effects should be played or not.

# pc_speaker_on = 1

# Volume music plays at (0 through 10)

# music_volume = 8

# Volume samples play at (0 through 10)

# sample_volume = 8

# Volume PC speaker SFX play at (0 through 10)

# pc_speaker_volume = 8

# The maximum number of samples that can be played simultaneously.
# Set to -1 for an unlimited number of samples.

# max_simultaneous_samples = -1


### Game options ###

# Name of the directory where MegaZeux should start.

# startup_path = /home/username/MegaZeux

# Name of world file that should be loaded when MZX starts.

# startup_file = caverns.mzx

# Default save file name given when you first try to save.

# save_file = saved.sav

# Speed MZX should start at (1 through 16)

# mzx_speed = 4

# Start in the editor by default

# startup_editor = 1

# Allow the use of editing cheats in regular gameplay. Set to 1
# to enable the cheats for either executable, "mzxrun" to enable
# them for MZXRun only, or to 0 to disable them for both. This
# feature is disabled by default.

# allow_cheats = 0

# Allow MegaZeux to automatically decrypt worlds without confirmation.
# Instead of replacing the world, the world will be decrypted either in
# memory or to a temporary file. Enabled by default.

# auto_decrypt_worlds = 1

# Enable standalone mode. This is a game mode meant to be used
# for packaged standalone MegaZeux games. In standalone mode:
# - ENTER_MENU, ESCAPE_MENU, HELP_MENU, F2_MENU and LOAD_MENU
#   can be used on the title screen to disable these menus.
# - HELP_MENU prevents the help menu from being opened even
#   when other builtin dialogs are open
# - EXIT_GAME can be used on the title screen to exit MegaZeux
# - The default title screen message does not appear
# - PLAY_GAME can be set to 1 to start a game
# - Using ALT-F4, closing the window or using other OS-specific
#   ways of closing an application will exit MZX entirely,
#   without asking for confirmation first
# Standalone mode can only be used from MZXRun.

# standalone_mode = 1

# Enable no titlescreen mode. This mode only works if standalone
# mode is enabled and causes the following:
# - The title screen is skipped and the game begins on the first
#   board.
# - Exiting to the title screen exits MegaZeux

# no_titlescreen = 1

### Board editor options ###

# Whether or not the spacebar can be used to toggle between
# your active buffer and the meta character "SPACE". If you
# set this, it will behave like older MegaZeux versions, but it
# will allow you to overwrite robots, for instance.
#
# If the setting is disabled (the default), replacing params
# will your new buffer will not require a toggle, and robots
# can only be deleted with the Delete key.

# editor_space_toggles = 1

# Whether or not the editor should center the viewport on the
# cursor in tab draw mode. Disabled by default; uncomment the
# following line to enable.

# editor_tab_focuses_view = 1

# Whether or not the editor will load board char sets and
# palettes when the board is changed. Disabled by default.

# editor_load_board_assets = 1

# Whether or not the thing menus (F3-F10) will place an object on
# the board when that object is selected. Enabled by default.

# editor_thing_menu_places = 1

# By default, the show thing keys (e.g. "Shift+F2: Show Robots") will block
# editing until a key or mouse button is pressed, after which they are turned
# off. If this option is set to 1, these keys will instead toggle on until
# the same key is pressed again, another show thing key is pressed, or until
# escape is pressed. This allows scrolling and other editing without having to
# press these keys repeatedly.

# editor_show_thing_toggles = 0

# This option controls the blinking speed when a show thing key is used.
# Lower numbers mean faster blinking; for example, when set to 4 (default),
# the character displayed by the blinking will alternate every 4 frames (~8 Hz).
# When set to 0, no blinking will occur and only the first character will display.

# editor_show_thing_blink_speed = 4

# If this is set to 1 (the default), enter can be used to split
# a line in the robot editor. However, the classic behaviour is
# that enter moves the cursor down a line. If you would like to
# restore the old behaviour, please set this to 0.

# editor_enter_splits = 1

# Set to 0 if you want the board editor to show the hotkey help
# by default.

# board_editor_hide_help = 1

# Set to 1 if you want the palette editor to hide the hotkey help
# when MegaZeux starts (default: don't hide).

# palette_editor_hide_help = 0

# The size of the board, overlay, vlayer, and char editor undo history
# stacks. Valid values are between 0 and 1000. Defaults to 100.

# undo_history_size = 100

# Defaults for new boards. Put these in [filename].editor.cnf to customize
# for each individual game, E.G. caverns.editor.cnf.  Viewport width/height
# must be set before offset, and board width/height must be set before any
# viewport vars.

# Options for the majority of these are 0 or 1, for off and on.
# Options for explosions_leave: space, ash, fire
# Options for saving: enabled, disabled, sensoronly
# Options for overlay: enabled, disabled, static, transparent

# board_default_width = 100
# board_default_height = 100
# board_default_viewport_w = 80
# board_default_viewport_h = 25
# board_default_viewport_x = 0
# board_default_viewport_y = 0
# board_default_can_shoot = 1
# board_default_can_bomb = 1
# board_default_fire_burns_spaces = 0
# board_default_fire_burns_fakes = 1
# board_default_fire_burns_trees = 1
# board_default_fire_burns_brown = 0
# board_default_fire_burns_forever = 0
# board_default_forest_to_floor = 0
# board_default_collect_bombs = 0
# board_default_restart_if_hurt = 0
# board_default_reset_on_entry = 0
# board_default_player_locked_ns = 0
# board_default_player_locked_ew = 0
# board_default_player_locked_att = 0
# board_default_time_limit = 0
# board_default_explosions_leave = ash
# board_default_saving = enabled
# board_default_overlay = enabled
# board_default_charset_path =
# board_default_palette_path =


### Robot editor options ###

# Enable or disable color coding in the editor.
# If color coding is off, ccode_string is used for the entire
# line.

# color_coding_on = 0

# Color codes for various types.
# For colors, use 255 to show a color box.

# ccode_chars = 14
# ccode_colors = 255
# ccode_commands = 15
# ccode_conditions = 15
# ccode_current_line = 11
# ccode_directions = 3
# ccode_equalities = 0
# ccode_extras = 7
# ccode_immediates = 10
# ccode_items = 11
# ccode_params = 2
# ccode_strings = 14
# ccode_things = 11

# Whether or not "extra words" should be displayed.
# For instance, if turned on, a line like "wait 1" will
# become "wait for 1."

# disassemble_extras = 0

# The base numbers should be displayed as. Valid options
# are 10 and 16 (decimal and hexidecimal). Hex numbers are
# prefixed with a dollar sign ($).

# disassemble_base = 16

# Changing this will allow invalid comments to automatically
# be marked to "delete" or "comment" instead of "ignore".
# Don't change this unless you know what you're doing.

# default_invalid_status = ignore

# Single line macros
# These are the traditional MZX macros; they may only
# occupy one line in their definition in this file. They
# correspond to parameter-less macros: when you press
# F6 through F10 the following strings are printed to
# the editor, verbatim. For instance if you set macro_1 to
# * "hello", * "hello" will be inserted every time you
# press F6 in the robot editor.
# The following are the default singe-line macros.

# macro_1 = char\s
# macro_2 = color\s
# macro_3 = goto\s
# macro_4 = send\s
# macro_5 = :\splayershot^

# Extended macros:
# See macro.txt.

# The following is an example extended macro.
# It sets up a sprite. If you find it useful, remove the #'s.

#macro_sprite = Sprite\sSetup
# (number255 spr_num)
# (number255 w, h, rx, ry)
# (number255 cx, cy, cw, ch)
# (number255 x, y)
# set spr!spr_num!_width !w!
# set spr!spr_num!_height !h!
# set spr!spr_num!_refx !rx!
# set spr!spr_num!_refy !ry!
# set spr!spr_num!_cwidth !cw!
# set spr!spr_num!_cheight !ch!
# set spr!spr_num!_cx !cx!
# set spr!spr_num!_cy !cy!
# put !x1! sprite p!#spr_num! !x! !y!

# Whether or not the palette should be reverted to default
# when the robot editor is load. If set, SMZX mode 0 will
# also be enforced.

# robot_editor_default_palette = 0

# Set to 0 if you want the robot editor to show the hotkey help
# and horizontal border by default.

# robot_editor_hide_help = 1


# Backup options

# Backups are save files made in the editor. They will only be made
# while in the main editor, not while in a sub-editor or a dialogue
# box of sorts. Note that this can cause slight pauses while editing,
# as the game saves.
# If you have multiple backups you should check the modified date in
# a console or file properties to determine which was most recently
# modified. The number will not indicate most recently saved.


CHANGELOG
2.51ak release:
+ Added Numerous new Counters
+ Added new inter robot targeting system
+ Added new SMZX mode, dynamic resolution setting
+ Added SMZX mode character editor, allowing better editing
+ Fixed fatal crash when taking a picture, also increased
  the number of picture you can take to 99
+ Removed Built-in ASCII charset
+ Added Random color start up
+ Fixed import World Feature

2.51s3.2 release:
+ Removed password protection (no use for it because of Getpw)
+ New counter: BIMesg. Set to 0 to turn off messages like "Ouch!"

2.51s3.1 release:
+ Fixed damage table, now works correctly
+ Fixed unworking alt+s
+ Fixed screenshot name
+ Fixed strange missile color when firing

2.51s3 release:

+ Fixed bug involving savegame name being overwritten by mod name
+ Fixed overlay transparency weirdness. Now, background of lower layer will
  ALWAYS show through if background of overlay is 0
+ Added in kev's refresh screen support. Press alt+w on the title screen or
  editor screen, and = in the game(don't ask, has to do with mzx's keybard
  handler).  Use this if your charset gets corrupted in windows.
+ Screen shot saving rotates file name (starts with screen0.pcx)
+ Fixed bug involving incorrect loading of max_health and others

2.51s2final release:

+ Deleted cpu checking code, was unnecesary and caused errors.
+ Fixed bug involving mod * settings being lost when loading
+ Should have fixed ems problems some were reporting, if not, just
  try to get more conventional memory.  Enclosed document (convmem.txt)
  should help you.
+ Fixed a bug involving direction checking.
+ Pressing ']' at most any time will now save a screenshot to screen.pcx
+ Pressing alt+8 in the editor will set the current mod to *.  Trust me.


2.5.1s2beta release:

+ added MOD "*", which allows a board to use whatever module the previous
  board was using; this necessitated a change in the .SAV file format
  (still need changes in the editor UI to access this feature outside of
   Robotic)
+ inmate's semantics for MOD "SOMETHIN.MOD*" aren't immediately possible,
  although a rough hack is in already
+ char edits are now not displayed until the beginning of the next cycle;
  the deferred display should eliminate a lot of flickering
+ an (at least partial) fix for the infamous UNDER bug (thanks JZig!)
+ MOUSEX, MOUSEY and BUTTONS are now buffered.
+ more sane magic handling
+ can load MZX 2.51 and 2.51S1 worlds, but will only save 2.5.1spider2 format


2.51S1 release:
 

+ Increased counters to 1000.
+ Added following counters: (ro) is read only, (r&w) is read + write
(r&w) MOUSEX, MOUSEY - Location of mouse cursor over the screen.
(ro)  BUTTONS - Status of the buttons (none=0,left=1,right=2,both=3)
(ro)  MBOARDX, MBOARDy - Location of mouse cursor over the board.
(ro)  SCROLLEDX, SCROLLEDY - Length the screen has been scrolled in each direction.
(ro)  PLAYERX, PLAYERY - Location of the player
(r&w) CURSORSTATE - Turns on & off the hardware cursor (0= off, 1= on, 0 default)

+ Added new independant counter like LOOPCOUNT, named LOCAL, it's now possible
  for a robot to search the screen or something without wasting a counter.
* New .MZX and .SAV formats for MZX2.51S1.
* .SAV files from 2.51S1 are incompatible with 2.51, and vice-versa
* 2.51S1 is capable of loading 2.51 .MZX world files.
* 2.51 is not capable of loading .MZX files created or opened & saved in 2.51S1




2.51 release: (quick bug-fix release while working on 3.00)

 * NOTE: I HAVE decided to release a 3.00 version of MegaZeux using
	PROTECTED MODE and a large number of enhancements. Visit the Software
	Visions web site for details. Betas will be PUBLIC.

 * Bug where moving north or west towards a board that is larger than 100
	tiles in a dimension may not work properly fixed.
 * Bugs in 2GDM.EXE preventing proper conversion of some S3Ms fixed.
 * A rare bug preventing internal MOD conversion fixed.
 * "Error opening MOD" will no longer appear during normal gameplay.
 * Exporting ANSis now adds a color to the end returning text to normal
	grey. (so TYPEing them in DOS no longer screws up the DOS colors)

2.50 release:

 * FOR ALL YOU 2.07 USERS- THE MAJOR UPGRADE is the new sound code! Support
	for stereo, 16-bit, more cards, 4 sound effects channels, 32 music
	channels, higher quality, and faster sound code! Less bugs! More formats
	supported! All non-MOD files MUST be converted to GDM using the included
	2GDM.EXE program. Formats supported- MOD (up to 32 channels), WOW/OCT/NST,
	S3M, 669, and MTM.

 * Volume controls (overall and SAM) added to F2-Settings (saved in .CFG file)
 * SAMs play at 2x volume.
 * MTM support added to 2GDM.EXE. Other formats aren't added because they
	aren't needed (not used often enough) or in the case of XM, because they
	support too many advanced features that MZX does not support. Partial
	XM support MAY be added later.
 * Many minor bug fixes in 2GDM.EXE's conversion routines.
 * Internal MOD conversion (in MegaZeux) sped up considerably.
 * For all those that upgraded from 2.07, see 2.49g and 2.48b WHATS-NEW
	for more stuff, including a couple interesting new Robotic features

 * This is the LAST planned release of MegaZeux. I may release bug-fix
	upgrades but that is it. I am moving on to new projects. (I will still
	support MegaZeux, of course.)

2.49g release:

 * New Import function added- You can import an ANSi to any position on the
	board.
 * Full SAM sound effect support as in the original 2.07 version (but using
	the new MOD code for up to 4 simultaneous channels)
 * The OPEN Robotic command will now push the Robot out of the way if the
	door would be blocked by the robot.
 * REL PLAYER and REL COUNTERS now affects the THISX/THISY counters
	(giving the distance FROM the player/counters TO the Robot) and the
	prefixes also affect the IF [dir] BLOCKED command, allowing you to

src/settings.c
/* MegaZeux
 *
 * Copyright (C) 1996 Greg Janson
 * Copyright (C) 1999 Charles Goetzman
 * Copyright (C) 2004 Gilead Kutnick <exophase@adelphia.net>
 * Copyright (C) 2018 Alice Rowan <petrifiedrowan@gmail.com>
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 2 of
 * the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
 */

#include "configure.h"
#include "core.h"
#include "error.h"
#include "event.h"
#include "game.h"
#include "graphics.h"
#include "settings.h"
#include "util.h"
#include "window.h"
#include "world_struct.h"

#include "audio/audio.h"

#include <string.h>

// Settings dialog-

//----------------------------
//
//    Speed- [   NN][^][v]
//
//   ( ) Digitized music on
//   ( ) Digitized music off
//
//   ( ) PC speaker SFX on
//   ( ) PC speaker SFX off
//
//  Sound card volumes-       //
//  Overall volume- 12345678  //
//  SoundFX volume- 12345678  //
//  PC Speaker SFX- 12345678  //
//
//      [OK]     [Cancel]
//
//----------------------------//

//  [OK]  [Cancel]  [Shader]  //

enum
{
  RESULT_CANCEL,
  RESULT_CONFIRM,
  RESULT_KEEP_OPEN,
  RESULT_SET_VIDEO_OUTPUT,
  RESULT_SET_SHADER
};

#ifdef CONFIG_RENDER_GL_PROGRAM
// Note- we search for .frags, then load the matching .vert too if present.
static const char *shader_exts[] = { ".frag", NULL };
static char shader_default_text[20];
static char shader_path[MAX_PATH];
static char shader_name[MAX_PATH];
static boolean is_glsl = false;
#endif

static int choose_video_output(struct world *mzx_world, const char **vo,
 int num_vo, int current_vo)
{
  int retval;
  struct dialog di;
  struct element *elements[] =
  {
    construct_list_box(2, 1, vo, num_vo, 12, 20, 0, &current_vo, NULL, false)
  };

  construct_dialog(&di, "Choose renderer...", 28, 5, 24, 14,
   elements, ARRAY_SIZE(elements), 0);

  retval = run_dialog(mzx_world, &di);
  destruct_dialog(&di);

  if(!retval)
    return current_vo;
  else
    return -1;
}

void game_settings(struct world *mzx_world)
{
  struct board *src_board = mzx_world->current_board;
  int mzx_speed, music, pcs;
  int prev_music;
  int music_volume, sound_volume, pcs_volume;
  struct dialog di;
  int dialog_result;
  int y_offset;
  int y_offset_buttons;
  int window_height;
  int num_elements;
  int start_option = -1;

  int ok_pos;

  const char *radio_strings_1[2] =
  {
    "Music/sample SFX on", "Music/sample SFX off"
  };
  const char *radio_strings_2[2] =
  {
    "PC speaker SFX on", "PC speaker SFX off"
  };
  struct element *elements[11];

  struct config_info *conf = get_config();

  const char *available_video_outputs[32];
  int num_available_video_outputs = 0;
  int cur_video_output = get_current_video_output();

  num_available_video_outputs =
   get_available_video_output_list(available_video_outputs,
    ARRAY_SIZE(available_video_outputs));

  mzx_speed = mzx_world->mzx_speed;
  music = !audio_get_music_on();
  pcs = !audio_get_pcs_on();
  music_volume = audio_get_music_volume();
  sound_volume = audio_get_sound_volume();
  pcs_volume = audio_get_pcs_volume();
  prev_music = !music;

  // Prevent previous keys from carrying through.
  force_release_all_keys();

  set_context(CTX_CONFIGURE);

  do
  {
    ok_pos = 6;
    num_elements = 8;
    y_offset = 0;
    y_offset_buttons = 0;

    if(!mzx_world->lock_speed)
    {
      y_offset += 2;
      y_offset_buttons += 2;
      num_elements++;
    }

#if !defined(CONFIG_WII) && !defined(CONFIG_SWITCH) && \
 !defined(__EMSCRIPTEN__) && !defined(CONFIG_PSVITA)
    // Emscripten's SDL port crashes on re-entry attempts.
    // Wii has multiple renderers but shouldn't display this option.
    // FIXME this is a hack. Fix the Wii renderers so they're switchable.
    // FIXME: Switch has undiagnosed crash bugs related to renderer switching.
    if(num_available_video_outputs > 1)
    {
      elements[ok_pos] = construct_button(4, 13 + y_offset_buttons,
       " Select renderer... ", RESULT_SET_VIDEO_OUTPUT);

      ok_pos++;
      num_elements++;
      y_offset_buttons += 2;
    }
#endif

#ifdef CONFIG_RENDER_GL_PROGRAM
    {
      const char *cur_output_name = conf->video_output;

      if(num_available_video_outputs)
        cur_output_name = available_video_outputs[cur_video_output];

      is_glsl = strstr(cur_output_name, "glsl") ? true : false;
    }

    if(is_glsl)
    {
      if(!shader_default_text[0])
      {
        if(conf->gl_scaling_shader[0])
          snprintf(shader_default_text, 20, "<conf: %.11s>",
           conf->gl_scaling_shader);

        else
          strcpy(shader_default_text, "<default: semisoft>");
      }
      strcpy(shader_path, mzx_res_get_by_id(GLSL_SHADER_SCALER_DIRECTORY));

      elements[ok_pos] = construct_file_selector(3, 13 + y_offset_buttons,
       "Scaling shader-", "Choose a scaling shader...", shader_exts,
       shader_default_text, 21, 0, shader_path, shader_name, RESULT_SET_SHADER);

      ok_pos++;
      num_elements++;
      y_offset_buttons += 3;
    }
#endif

    if(!mzx_world->lock_speed)
    {
      elements[ok_pos + 2] = construct_number_box(2, 2, "Speed- ", 1, 16,
       NUMBER_SLIDER, &mzx_speed);

      // Start at the Speed setting if enabled.
      if(start_option < 0)
        start_option = ok_pos + 2;
    }

    // Start at the music/samples toggle by default.
    if(start_option < 0)
      start_option = 0;

    elements[0] = construct_radio_button(4, 2 + y_offset,
     radio_strings_1, 2, 20, &music);
    elements[1] = construct_radio_button(4, 5 + y_offset,
     radio_strings_2, 2, 18, &pcs);
    elements[2] = construct_label(2, 8 + y_offset,
     "Audio volumes-");
    elements[3] = construct_number_box(7, 9 + y_offset,
     "Music- ", 0, 10, NUMBER_SLIDER, &music_volume);
    elements[4] = construct_number_box(5, 10 + y_offset,
     "Samples- ", 0, 10, NUMBER_SLIDER, &sound_volume);
    elements[5] = construct_number_box(2, 11 + y_offset,
     "PC Speaker- ", 0, 10, NUMBER_SLIDER, &pcs_volume);

    elements[ok_pos] = construct_button(7, 13 + y_offset_buttons,
     "OK", RESULT_CONFIRM);

    elements[ok_pos + 1] = construct_button(15, 13 + y_offset_buttons,
     "Cancel", RESULT_CANCEL);

    window_height = (16 + y_offset_buttons);
    construct_dialog(&di, "Game Settings", 25, (25 - window_height) / 2,
     30, window_height, elements, num_elements, start_option);

    dialog_result = run_dialog(mzx_world, &di);
    start_option = di.current_element;

    // Prevent UI keys from carrying through.
    force_release_all_keys();

    if(dialog_result == RESULT_CONFIRM)
    {
      audio_set_pcs_volume(pcs_volume);

      if(music_volume != audio_get_music_volume())
      {
        audio_set_music_volume(music_volume);

        if(mzx_world->active)
          audio_set_module_volume(src_board->volume);
      }

      if(sound_volume != audio_get_sound_volume())
        audio_set_sound_volume(sound_volume);

      music ^= 1;
      pcs ^= 1;

      audio_set_pcs_on(pcs);
      audio_set_music_on(music);

      if(!music && prev_music && mzx_world->active)
      {
        // Turn off music and sound effects.
        audio_end_module();
        audio_end_sample();
      }
      else

      if(music && !prev_music && mzx_world->active)
      {
        // Turn on music.
        load_game_module(mzx_world, mzx_world->real_mod_playing, false);
      }

      mzx_world->mzx_speed = mzx_speed;
    }
    else

    if(dialog_result == RESULT_SET_VIDEO_OUTPUT)
    {
      int new_video_output = choose_video_output(mzx_world,
       available_video_outputs, num_available_video_outputs, cur_video_output);

      if(new_video_output >= 0)
      {
        if(!change_video_output(conf, available_video_outputs[new_video_output]))
          error("Failed to set renderer.", ERROR_T_ERROR, ERROR_OPT_OK, 0x6000);

        cur_video_output = get_current_video_output();

        // HACK: after creating a new X11 window SDL seems to like to send the
        // previous frame's events again sometimes, so just flush the events
        update_event_status();
      }
    }

#ifdef CONFIG_RENDER_GL_PROGRAM
    else

    // Shader selection
    if(is_glsl && (dialog_result == RESULT_SET_SHADER) && shader_name[0])
    {
      size_t offset = strlen(shader_path) + 1;
      boolean shader_res;
      char *pos;

      if(strlen(shader_name) > offset)
      {
        pos = strrchr(shader_name + offset, '.');

        if(pos) *pos = 0;
        shader_res = switch_shader(shader_name + offset);
        if(pos) *pos = '.';

        if(!shader_res)
        {
          // If the shader failed to load, the default should have been loaded.
          strcpy(shader_default_text, "<default: semisoft>");
          shader_name[0] = 0;
        }
      }
    }
#endif
    destruct_dialog(&di);
  }
  while(dialog_result >= RESULT_KEEP_OPEN);

  pop_context();
}
